File: task2/task2.py
import sys

def main():
    print("Введите количество пар городов: ", end="")
    n = int(input())
    pairs = []
    for _ in range(n):
        a, b = input().split()
        pairs.append((a, b))

    error = False

    for p in pairs:
        if p[0] == p[1]:
            error = True
            break

    start_count = {}
    for p in pairs:
        start_count[p[0]] = start_count.get(p[0], 0) + 1
        if start_count[p[0]] > 1:
            error = True
            break

    end_count = {}
    for p in pairs:
        end_count[p[1]] = end_count.get(p[1], 0) + 1
        if end_count[p[1]] > 1:
            error = True
            break

    if error:
        print("Ошибка")
        return

    next_map = {p[0]: p[1] for p in pairs}

    starts = set()
    for p in pairs:
        if p[0] not in end_count:
            starts.add(p[0])

    if not starts:
        print("Ошибка")
        return

    best_end = ""
    max_len = -1

    for start in starts:
        cur = start
        length = 0
        while cur in next_map:
            cur = next_map[cur]
            length += 1
        if length > max_len:
            max_len = length
            best_end = cur

    print(best_end)


if __name__ == "__main__":
    main()

File: task2/task2.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in, "UTF-8");
        OutputStream out = System.out;
        
        out.write("Введите количество пар городов: ".getBytes("UTF-8"));
        int n = sc.nextInt();
        
        List<String[]> pairs = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            String a = sc.next();
            String b = sc.next();
            pairs.add(new String[]{a, b});
        }

        boolean error = false;

        for (String[] p : pairs) {
            if (p[0].equals(p[1])) {
                error = true;
                break;
            }
        }

        Map<String, Integer> startCount = new HashMap<>();
        for (String[] p : pairs) {
            int cnt = startCount.getOrDefault(p[0], 0) + 1;
            if (cnt > 1) {
                error = true;
                break;
            }
            startCount.put(p[0], cnt);
        }

        Map<String, Integer> endCount = new HashMap<>();
        for (String[] p : pairs) {
            int cnt = endCount.getOrDefault(p[1], 0) + 1;
            if (cnt > 1) {
                error = true;
                break;
            }
            endCount.put(p[1], cnt);
        }

        if (error) {
            out.write("Ошибка".getBytes("UTF-8"));
            out.write('\n');
            return;
        }

        Map<String, String> next = new HashMap<>();
        for (String[] p : pairs) {
            next.put(p[0], p[1]);
        }

        Set<String> starts = new HashSet<>();
        for (String[] p : pairs) {
            if (!endCount.containsKey(p[0])) {
                starts.add(p[0]);
            }
        }

        if (starts.isEmpty()) {
            out.write("Ошибка".getBytes("UTF-8"));
            out.write('\n');
            return;
        }

        String bestEnd = "";
        int maxLen = -1;
        for (String start : starts) {
            String cur = start;
            int len = 0;
            while (next.containsKey(cur)) {
                cur = next.get(cur);
                len++;
            }
            if (len > maxLen) {
                maxLen = len;
                bestEnd = cur;
            }
        }

        out.write(bestEnd.getBytes("UTF-8"));
        out.write('\n');
    }
}